### PR TITLE
don't crash when requesting color for the NaN value (fix #29509)

### LINKED ISSUE #29509: QGIS 3.4.6 crashed when I tried classifying a layer
Author Name: **Raphael Dumas** (@radumas)
Original Redmine Issue: [21693](https://issues.qgis.org/issues/21693)
Affected QGIS version: 3.6.0
Redmine category:symbology

---
## User Feedback

I clicked "Classify" trying to style a line layer with 8 features by "categorized" with a unique colour for each feature. The layer was from a custom query from PostGIS. 

## Report Details

*Crash ID*: ac16e11417d9f0b521055c760cd79e2788f914f6

*Stack Trace*
```
QgsColorBrewerColorRamp::color :
QgsCategorizedSymbolRenderer::updateColorRamp :
QgsCategorizedSymbolRendererWidget::addCategories :
QMetaObject::activate :
QAbstractButton::clicked :
QAbstractButton::click :
QAbstractButton::mouseReleaseEvent :
QWidget::event :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QApplicationPrivate::sendMouseEvent :
QSizePolicy::QSizePolicy :
QSizePolicy::QSizePolicy :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QGuiApplicationPrivate::processMouseEvent :
QWindowSystemInterface::sendWindowSystemEvents :
QEventDispatcherWin32::processEvents :
TranslateMessageEx :
TranslateMessage :
QEventDispatcherWin32::processEvents :
qt_plugin_query_metadata :
QEventLoop::exec :
QDialog::exec :
QgisApp::showLayerProperties :
QgisApp::layerTreeViewDoubleClicked :
QMetaObject::activate :
QAbstractItemView::doubleClicked :
QTreeView::mouseDoubleClickEvent :
QWidget::event :
QFrame::event :
QAbstractItemView::viewportEvent :
QCoreApplicationPrivate::sendThroughObjectEventFilters :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QApplicationPrivate::sendMouseEvent :
QSizePolicy::QSizePolicy :
QSizePolicy::QSizePolicy :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QGuiApplicationPrivate::processMouseEvent :
QWindowSystemInterface::sendWindowSystemEvents :
QEventDispatcherWin32::processEvents :
TranslateMessageEx :
TranslateMessage :
QEventDispatcherWin32::processEvents :
qt_plugin_query_metadata :
QEventLoop::exec :
QCoreApplication::exec :
main :
BaseThreadInitThunk :
RtlUserThreadStart :

```

*QGIS Info*
QGIS Version: 3.4.6-Madeira
QGIS code revision: cc0e375a9f
Compiled against Qt: 5.11.2
Running against Qt: 5.11.2
Compiled against GDAL: 2.4.1
Running against GDAL: 2.4.1

*System Info*
CPU Type: x86_64
Kernel Type: winnt
Kernel Version: 6.1.7601

---

- [classification_bug.qgz](https://issues.qgis.org/attachments/download/14684/classification_bug.qgz) (Raphael Dumas)
- [bluetooth_segments.geojson](https://issues.qgis.org/attachments/download/14685/bluetooth_segments.geojson) (Raphael Dumas)

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Prevents crash in the `QgsColorBrewerColorRamp` class when requesting color corresponding to the `NaN` value. 

Fixes #29509. Tests for color ramps added.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: tests/src/python/test_qgscolorramp.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        self.assertEqual(ramp.colors(), 13)
+
+        # color for value
+        ramp.setColors(5)
+        self.assertEqual(ramp.color(0).name(), "#d7191c")
+        self.assertEqual(ramp.color(1).name(), "#2b83ba")
+        self.assertEqual(ramp.color(0.5).name(), "#ffffbf")
+        self.assertEqual(ramp.color(2).name(), "#000000")
+        self.assertEqual(ramp.color(-1).name(), "#000000")
+        self.assertEqual(ramp.color(float('nan')).name(), "#000000")
```
Comment (nyalldawson) [inline tests/src/python/test_qgscolorramp.py]:
```suggestion
        self.assertFalse(ramp.color(float('nan')).isValid())
```

File: tests/src/python/CMakeLists.txt [hunk 2]
```diff
@@ -28,6 +28,7 @@ ADD_PYTHON_TEST(PyQgsBox3d test_qgsbox3d.py)
 ADD_PYTHON_TEST(PyQgsCategorizedSymbolRenderer test_qgscategorizedsymbolrenderer.py)
 ADD_PYTHON_TEST(PyQgsCheckableComboBox test_qgscheckablecombobox.py)
 ADD_PYTHON_TEST(PyQgsColorButton test_qgscolorbutton.py)
+ADD_PYTHON_TEST(PyQgsColorRamp test_qgscolorramp.py)
```
Comment (nyalldawson) [inline tests/src/python/CMakeLists.txt]: There's already a test file containing lots of ramp tests as test_qgsvectorcolorramp.py -- your new tests should be merged into the existing file (and ideally renamed away from the outdated vectorcolorramp name)

Comment (alexbruy) [reply tests/src/python/CMakeLists.txt]: Done

File: tests/src/python/test_qgscolorramp.py (hunk 1 above)
Comment (alexbruy) [reply tests/src/python/test_qgscolorramp.py]: Thanks, fixed

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**